fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java

```diff
@@ -1118,6 +1132,55 @@ protected RuntimeException mapAsyncToSyncException(@Nonnull Throwable ex) {
         return future.get();
     }
 
+    /**
+     * Log warning and close tracked contexts that have been open for too long.
+     * @param minAgeSeconds number of seconds above which to warn
+     * @return number of such contexts found
+     */
+    public int warnAndCloseOldTrackedOpenContexts(long minAgeSeconds) {
+        long nanoTime = System.nanoTime() - TimeUnit.SECONDS.toNanos(minAgeSeconds);
+        if (trackedOpenContexts.isEmpty()) {
+            return 0;
+        }
+        try {
+            if (trackedOpenContexts.firstKey() > nanoTime) {
+                return 0;
+            }
+        } catch (NoSuchElementException ex) {
+            return 0;
+        }
+        int count = 0;
+        for (FDBRecordContext context : trackedOpenContexts.headMap(nanoTime, true).values()) {
+            KeyValueLogMessage msg = KeyValueLogMessage.build("context not closed",
+                    LogMessageKeys.AGE_SECONDS, TimeUnit.NANOSECONDS.toSeconds(nanoTime - context.getTrackOpenTimeNanos()),
+                    LogMessageKeys.TRANSACTION_ID, context.getTransactionId());
+            if (context.getOpenStackTrace() != null) {
+                LOGGER.warn(msg.toString(), context.getOpenStackTrace());
+            } else {
+                LOGGER.warn(msg.toString());
+            }
+            context.close();
+            count++;
+        }
+        return count;
+    }
+
+    protected void trackOpenContext(FDBRecordContext context) {
+        long key = System.nanoTime();
+        while (key == 0 || trackedOpenContexts.putIfAbsent(key, context) != null) {
+            key++;  // Might not have nanosecond resolution and need something non-zero and unique.
```

Comment:
Just an observation: I wasn't aware that `nanoTime()` could return 0. That's useful to be aware of.